### PR TITLE
Fix #601 Stop considering properties in anonymous class as local variables

### LINKED ISSUE #601: False-positive message "Avoid unused local variables" in anonymous class
There is false-positive message "Avoid unused local variables such as '$prop'" in anonymous class placed inside another class or function:
```
function getData($var)
{
    return new class($var)
    {
        private $prop;

        public function __construct($var)
        {
            $this->prop = $var;
        }

        public function getProp()
        {
            return $this->prop;
        }
    };
}

$data = getData('123');
echo $data->getProp();
```
Message: "Avoid unused local variables such as '$prop'."
Version: 2.6.0

### PR DESCRIPTION
**Pull request by kylekatarnls**
Type: bugfix
Issue: Fix #601 ([confirmed](https://github.com/phpmd/phpmd/issues/601#issuecomment-630183576))
Breaking change: no

### CODE REVIEW

**Comment by ravage84** [comment]
@kylekatarnls for the future, can we please not chnage code AND refactor the tests?
Doing so unnecessarily increases the size of PR to review and also increases the risk of missing an introduced bug.

File: src/test/php/PHPMD/Rule/UnusedLocalVariableTest.php [hunk 1]
```diff
@@ -28,650 +28,47 @@
...
-     * @return void
-     */
-    public function testRuleDoesNotApplyToRequestSuperGlobal()
-    {
-        $rule = new UnusedLocalVariable();
-        $rule->addProperty('allow-unused-foreach-variables', 'false');
-        $rule->setReport($this->getReportWithNoViolation());
-        $rule->apply($this->getMethod());
-    }
+        if (preg_match('/Whitelisted/', $file)) {
```
Comment (ravage84) [inline src/test/php/PHPMD/Rule/UnusedLocalVariableTest.php:44]: I think this should have a comment explaining it.

File: src/test/php/PHPMD/Rule/UnusedLocalVariableTest.php [hunk 2]
```diff
@@ -28,650 +28,47 @@
...
      */
-    public function testRuleDoesNotApplyToPostSuperGlobal()
+    public function getRule($file)
     {
         $rule = new UnusedLocalVariable();
-        $rule->addProperty('allow-unused-foreach-variables', 'false');
-        $rule->setReport($this->getReportWithNoViolation());
-        $rule->apply($this->getMethod());
-    }
+        $rule->addProperty(
```
Comment (ravage84) [inline src/test/php/PHPMD/Rule/UnusedLocalVariableTest.php:39]: I think this should have a comment explaining it. And may be extract the ``preg_match()`` part in a self-explaining intermediate variable.

File: src/test/php/PHPMD/AbstractTest.php [hunk 3]
```diff
@@ -187,22 +188,49 @@ protected function getFunction()
...
+     * @return AbstractNode
+     */
+    protected function getNodeForTestFile($file)
+    {
+        $source = $this->parseSource($file);
+        $class = $source
+            ->getTypes()
+            ->current();
+
+        return $class
```
Comment (ravage84) [inline src/test/php/PHPMD/AbstractTest.php:219]: A ternary spreading 13 lines? I don't think we are a good example here...

File: src/test/php/PHPMD/AbstractTest.php [hunk 4]
```diff
@@ -213,7 +241,17 @@ protected function getMethodNodeForTestFile($file)
     protected function expectRuleHasViolationsForFile(Rule $rule, $expectedInvokes, $file)
     {
         $rule->setReport($this->getReportMock($expectedInvokes));
-        $rule->apply($this->getMethodNodeForTestFile($file));
+
+        try {
+            $rule->apply($this->getNodeForTestFile($file));
+        } catch (PHPUnit_Framework_ExpectationFailedException $failedException) {
```
Comment (ravage84) [inline src/test/php/PHPMD/AbstractTest.php:247]: What is the reason for this?

File: src/main/php/PHPMD/Rule/AbstractLocalVariable.php [hunk 5]
```diff
@@ -220,6 +207,7 @@ protected function getVariableImage($variable)
      * Or return the input as is if it's not an ASTNode PHPMD node.
      *
      * @param mixed $node
+     *
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/AbstractLocalVariable.php:210]: Please remove.

File: src/test/php/PHPMD/AbstractTest.php [hunk 6]
```diff
@@ -187,22 +188,49 @@ protected function getFunction()
...
         );
     }
 
+    /**
+     * Get the node for a given test file.
+     *
+     * This method can be overridden to select the node to test with the rule.
+     *
+     * @param string $file
+     *
```
Comment (ravage84) [inline src/test/php/PHPMD/AbstractTest.php:209]: Please remove line.

File: src/test/php/PHPMD/AbstractTest.php [hunk 7]
```diff
@@ -187,22 +188,49 @@ protected function getFunction()
      * Returns the first method as a MethodNode for a given test file.
      *
      * @param string $file
-     * @return MethodNode
+     * @return ClassNode
```
Comment (ravage84) [inline src/test/php/PHPMD/AbstractTest.php:191]: This is totally out of scope of this PR. :disappointed:

**Review by ravage84 — DISMISSED** [review]
Thank you for the fix. But Mashing three separte things into one PR makes me sad.

1. Actual fix (which looks fine)
2. Refactoring the ttest class to the new simpler style
3. Changing some inner workings of the abstract test.

I'd prefer if we could split that into three or at least two PRs.

File: src/test/php/PHPMD/AbstractTest.php (hunk 4 above)
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractTest.php:247]: Adding `basename($file)."\n".` at the begin of the exception message. See below.

File: src/test/php/PHPMD/AbstractTest.php (hunk 7 above)
Comment (kylekatarnls) [reply src/test/php/PHPMD/AbstractTest.php:191]: I will group this in a PR dedicated to AbstractTest update.

**Comment by kylekatarnls** [comment]
@ravage84 Please don't commit on this PR for the testing changes, I handled them in #787 as you requested.

File: src/test/php/PHPMD/Rule/UnusedLocalVariableTest.php (hunk 1 above)
Comment (kylekatarnls) [reply src/test/php/PHPMD/Rule/UnusedLocalVariableTest.php:49]: Done.

File: src/test/php/PHPMD/Rule/UnusedLocalVariableTest.php (hunk 2 above)
Comment (kylekatarnls) [reply src/test/php/PHPMD/Rule/UnusedLocalVariableTest.php:42]: Done.

**Review by tvbeek — APPROVED** [review]

File: src/main/php/PHPMD/Rule/AbstractLocalVariable.php [hunk 8]
```diff
@@ -325,4 +311,31 @@ protected function isPassedByReference($variable)
 
         return isset($parameters[$argumentPosition]) && $parameters[$argumentPosition]->isPassedByReference();
     }
+
+    /**
+     * Prepend "::" if the variable has a ASTMemberPrimaryPrefix.
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/AbstractLocalVariable.php:316]: Can we document why we are doing this?
In the sense of "document not (only) what you are doing, but why you are doing it".

File: src/main/php/PHPMD/Rule/AbstractLocalVariable.php [hunk 9]
```diff
@@ -204,30 +205,15 @@ protected function getVariableImage($variable)
...
-        $base = $variable;
-        $parent = $this->getNode($variable->getParent());
-
-        while ($parent instanceof ASTArrayIndexExpression &&
-            $base instanceof ASTNode &&
-            $parent->getChild(0) === $base->getNode()
-        ) {
-            $base = $parent;
-            $parent = $this->getNode($base->getParent());
+        if (substr($image, 0, 1) === '$' && $this->getNode($variable->getParent()) instanceof ASTFieldDeclaration) {
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/AbstractLocalVariable.php:208]: Could we extract this to a protected method, too and document the intention?

File: src/main/php/PHPMD/Rule/UnusedLocalVariable.php [hunk 10]
```diff
@@ -101,7 +101,7 @@ protected function removeParameters(AbstractCallableNode $node)
         $declarators = $parameters->findChildrenOfType('VariableDeclarator');
 
         foreach ($declarators as $declarator) {
-            unset($this->images[$declarator->getImage()]);
+            unset($this->images[$this->getVariableImage($declarator)]);
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/UnusedLocalVariable.php:104]: I like this cleanup of redundancy 👍

File: src/main/php/PHPMD/Rule/UnusedLocalVariable.php [hunk 11]
```diff
@@ -199,7 +198,7 @@ protected function storeImage($imageName, ASTNode $node)
      */
     protected function collectLiteral(ASTNode $node)
     {
-        $variable = '$' . trim($node->getImage(), '\'\"');
```
Comment (ravage84) [inline src/main/php/PHPMD/Rule/UnusedLocalVariable.php:202]: This never worked, I presume?

**Review by ravage84 — COMMENTED** [review]
Haven't looked into it too deeply. Would be great if we could address my notes but apart from that, go with it.

File: src/main/php/PHPMD/Rule/AbstractLocalVariable.php (hunk 9 above)
Comment (kylekatarnls) [reply src/main/php/PHPMD/Rule/AbstractLocalVariable.php]: Done.

File: src/main/php/PHPMD/Rule/AbstractLocalVariable.php (hunk 8 above)
Comment (kylekatarnls) [reply src/main/php/PHPMD/Rule/AbstractLocalVariable.php:317]: Added.

**Review by tvbeek — APPROVED** [review]
